<launch>
  <arg name="mav_name" default="$(env HOSTNAME)"/>
  <arg name="output" default="screen"/>
  <arg name="imu_frame_id" default="adis16448bmlz"/>
  <arg name="lidar_lite_frame_id" default="lidar_lite"/>
  <arg name="us_d1_frame_id" default="us_d1"/>

  <group ns="$(arg mav_name)">
    <node name="rosserial_server" pkg="rosserial_server" type="serial_node" args="_port:=/dev/versavis" respawn="true" output="$(arg output)"/>

    <node name="baro_transform_node" pkg="versavis" type="baro_transform_node" respawn="true" output="$(arg output)">
      <remap from="in" to="$(arg imu_frame_id)/baro_micro"/>
      <remap from="out" to="$(arg imu_frame_id)/baro"/>
      <param name="frame_id" type="string" value="$(arg imu_frame_id)"/>
      <param name="var" type="double" value="-1.0"/>
      <param name="scale" type="double" value="$(eval 0.02 * 1.0e2)"/>
    </node>

    <node name="imu_transform_node" pkg="versavis" type="imu_transform_node" respawn="true" output="$(arg output)">
      <remap from="in" to="$(arg imu_frame_id)/imu_micro"/>
      <remap from="out" to="$(arg imu_frame_id)/imu"/>
      <param name="frame_id" type="string" value="$(arg imu_frame_id)"/>
      <param name="gyro_cov" type="double" value="4.0e-06"/>
      <param name="acc_cov" type="double" value="0.0002"/>
      <param name="acc_scale" type="double" value="$(eval 9.807 / 1200.0)"/>
      <param name="gyro_scale" type="double" value="$(eval 0.04 * 3.14159265359 / 180.0)"/>
    </node>

    <node name="mag_transform_node" pkg="versavis" type="mag_transform_node" respawn="true" output="$(arg output)">
      <remap from="in" to="$(arg imu_frame_id)/mag_micro"/>
      <remap from="out" to="$(arg imu_frame_id)/mag"/>
      <param name="frame_id" type="string" value="$(arg imu_frame_id)"/>
      <param name="var" type="double" value="-1.0"/>
      <param name="scale" type="double" value="$(eval 0.0001429 * 1.0e-4)"/>
    </node>

    <node name="temp_transform_node" pkg="versavis" type="temp_transform_node" respawn="true" output="$(arg output)">
      <remap from="in" to="$(arg imu_frame_id)/temp_micro"/>
      <remap from="out" to="$(arg imu_frame_id)/temp"/>
      <param name="frame_id" type="string" value="$(arg imu_frame_id)"/>
      <param name="var" type="double" value="-1.0"/>
      <param name="scale" type="double" value="0.07386"/>
      <param name="offset" type="double" value="31.0"/>
    </node>

    <!-- https://static.garmin.com/pumac/LIDAR_Lite_v3_Operation_Manual_and_Technical_Specifications.pdf -->
    <node name="lidar_lite_transform_node" pkg="versavis" type="lidar_lite_transform_node" respawn="true" output="$(arg output)">
      <remap from="in" to="$(arg lidar_lite_frame_id)/data_micro"/>
      <remap from="out" to="$(arg lidar_lite_frame_id)/data"/>
      <param name="frame_id" type="string" value="$(arg lidar_lite_frame_id)"/>
      <param name="fov" type="double" value="0.008"/>
      <param name="min_range" type="double" value="0.05"/>
      <param name="max_range" type="double" value="30.0"/>
      <param name="var_low" type="double" value="$(eval (0.025 / 3.0)**2)"/>
      <param name="var_high" type="double" value="$(eval (0.1 / 3.0)**2)"/>
      <param name="scale" type="double" value="0.01"/>
      <param name="close_range" type="double" value="5.0"/>
    </node>

    <!-- https://cdn.shopify.com/s/files/1/0113/0414/0900/files/User_Manual_US-D1.pdf?16288212927919010227 -->
    <node name="us_d1_transform_node" pkg="versavis" type="us_d1_transform_node" respawn="true" output="$(arg output)">
      <remap from="in" to="$(arg us_d1_frame_id)/data_micro"/>
      <remap from="out" to="$(arg us_d1_frame_id)/data"/>
      <param name="frame_id" type="string" value="$(arg us_d1_frame_id)"/>
      <param name="fov" type="double" value="0.35"/>
      <param name="min_range" type="double" value="0.5"/>
      <param name="max_range" type="double" value="50.0"/>
      <param name="var_low" type="double" value="$(eval (0.06 / 3.0)**2)"/>
      <param name="var_high" type="double" value="$(eval (0.04 / 3.0)**2)"/>
      <param name="scale" type="double" value="0.01"/>
      <param name="close_range" type="double" value="1.0"/>
    </node>

  </group>

</launch>
